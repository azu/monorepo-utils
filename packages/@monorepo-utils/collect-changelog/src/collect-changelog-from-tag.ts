// MIT Â© 2017 azu
"use strict";
import * as path from "path";
import * as fs from "fs";
import { getPackages } from "@monorepo-utils/package-utils";

const parse = require("cclog-parser");

interface CclogObject {
    versions: string[];
    changes: ChangesByVersion;
}

interface ChangesByVersion {
    [index: string]: Changes;
}

export type Changes = {
    fixes?: string[];
    features?: string[];
    breakingChanges?: string[];
};

export interface ChangLogResult {
    name: string;
    version: string;
    changes: Changes;
    // if has changes at least one, set true
    hasChanges: boolean;
}

function getChangelog(filePath: string) {
    return new Promise<CclogObject>((resolve, reject) => {
        if (!fs.existsSync(filePath)) {
            return reject(new Error(`${filePath} is not found`));
        }
        const content = fs.readFileSync(filePath, "utf-8");
        resolve(parse(content));
    });
}

function getLog(changelogFilePath: string, { name, version }: { name: string; version: string }): Promise<Changes> {
    return getChangelog(changelogFilePath).then((result: CclogObject) => {
        if (!result.changes[version]) {
            throw new Error(`Not found that version: ${name}@${version}`);
        }
        return result.changes[version];
    });
}

function splitLernaTag(lernaTag: string) {
    const splitIndex = lernaTag.lastIndexOf("@");
    return {
        name: lernaTag.slice(0, splitIndex),
        version: lernaTag.slice(splitIndex + 1)
    };
}

function filterEmpty(array?: string[]) {
    if (!array) {
        return [];
    }
    return array.filter(item => item.trim().length !== 0);
}

function filter(changes: Changes): Changes {
    if (filterEmpty(changes.breakingChanges).length === 0) {
        delete changes.breakingChanges;
    }
    if (filterEmpty(changes.features).length === 0) {
        delete changes.features;
    }
    if (filterEmpty(changes.fixes).length === 0) {
        delete changes.fixes;
    }
    return changes;
}

export function convertCClogToResult(name: string, version: string, changes: Changes): ChangLogResult {
    const filteredChanges = filter(changes);
    return {
        name,
        version,
        changes: filteredChanges,
        hasChanges: Object.keys(filteredChanges).length > 0
    };
}

/**
 * Find changelog and return
 * @param projectRootDirectory project root directory that has lerna.json or package.json
 * @param lernaTag package@version string
 */
export function findChangelog(projectRootDirectory: string, lernaTag: string): Promise<ChangLogResult> {
    const meta = splitLernaTag(lernaTag);
    if (!meta.name || !meta.version) {
        throw new Error("It is not lerna tag" + lernaTag);
    }
    const matchPackage = getPackages(projectRootDirectory).find(value => {
        return value.packageJSON.name === meta.name;
    });
    if (!matchPackage) {
        throw new Error(`Not found match package: ${meta.name}@${meta.version}`);
    }
    return getLog(`${path.join(matchPackage.location)}/CHANGELOG.md`, meta).then(changes => {
        return convertCClogToResult(meta.name, meta.version, changes);
    });
}
