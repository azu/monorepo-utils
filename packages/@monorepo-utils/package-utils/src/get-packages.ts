"use strict";
// Based on https://github.com/azz/get-monorepo-packages
// MIT (c) 2018 Lucas Azzola
const globby = require("globby");
import loadJsonFile = require("load-json-file");
import * as path from "upath";
import * as fs from "fs";
import * as yaml from "yaml";

const loadPackage = (packagePath: string): object => {
    const pkgJsonPath = path.join(packagePath, "package.json");
    if (fs.existsSync(pkgJsonPath)) {
        return loadJsonFile.sync(pkgJsonPath);
    }
    return {};
};

export const findPackages = (packageSpecs: string[], rootDirectory: string): PackageResult[] => {
    return packageSpecs
        .reduce((pkgDirs, pkgGlob) => {
            const target = path.join(rootDirectory, pkgGlob);
            const filePathList = globby.hasMagic(pkgGlob)
                ? globby.sync(target, {
                      onlyFiles: false,
                  })
                : [target];
            return pkgDirs.concat(filePathList);
        }, [] as string[])
        .map((location: string) => {
            return { location, packageJSON: loadPackage(location) };
        })
        .filter((res) => {
            return res.packageJSON && "name" in res.packageJSON;
        });
};

export interface PackageResult {
    // path to package's directory
    location: string;
    // package.json content
    packageJSON: any;
}

export const getPackages = (rootDirectory: string): PackageResult[] => {
    const lernaJsonPath = path.join(rootDirectory, "lerna.json");
    // prefer to use yarn workspace instead of lerna.json
    if (fs.existsSync(lernaJsonPath)) {
        const lernaJson = loadJsonFile.sync(lernaJsonPath);
        const hasNotWorkspacesInLernaJson = (v: any): v is { packages: string[] } => {
            return !("useWorkspaces" in v);
        };
        if (hasNotWorkspacesInLernaJson(lernaJson)) {
            return findPackages(lernaJson.packages, rootDirectory);
        }
    }

    const pkgJsonPath = path.join(rootDirectory, "package.json");
    if (fs.existsSync(pkgJsonPath)) {
        const pkgJson = loadJsonFile.sync(pkgJsonPath);
        const hasWorkspacesInPackageJson = (
            v: any
        ): v is {
            workspaces: { packages: string[] } | string[];
        } => {
            return "workspaces" in v;
        };
        if (hasWorkspacesInPackageJson(pkgJson)) {
            // "workspaces": []
            if (Array.isArray(pkgJson.workspaces)) {
                return findPackages(pkgJson.workspaces, rootDirectory);
            } else if (Array.isArray(pkgJson.workspaces.packages)) {
                // "workspaces": { "packages": [] }
                return findPackages(pkgJson.workspaces.packages, rootDirectory);
            }
        }
    }

    const pnpmYamlPath = path.join(rootDirectory, "pnpm-workspace.yaml");
    if (fs.existsSync(pnpmYamlPath)) {
        const contents = fs.readFileSync(pnpmYamlPath, "utf8");
        const pnpmConfig = yaml.parse(contents);

        if ("packages" in pnpmConfig && Array.isArray(pnpmConfig.packages)) {
            return findPackages(pnpmConfig.packages, rootDirectory);
        }
    }

    // Bail if we don't find any packages
    return [];
};
