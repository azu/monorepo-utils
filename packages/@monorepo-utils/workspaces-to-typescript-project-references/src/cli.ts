import meow from "meow";
import path from "upath";
import { toProjectReferences, DEFAULT_TSCONFIGPATH } from "./index";

export const cli = meow(
    `
    Usage
      $ workspaces-to-typescript-project-references
 

    Options
      --root             [Path:string] Root directory of the monorepo. 
                         Default: current working directory

      --check            If set the flag, check only differences of tsconfig.json and does not update tsconfig.json.
                         If the check is failed, exit status 1. It is useful for testing.
       
      --plugin           [Path:string] Path to plugin script.
                         Load the plugin script as module and use it. 
                           
      --tsconfigPath     [Path:string] Use alternative config path inside the package. e.g.: tsconfig.test.json
                         Default: tsconfig.json

    Examples
      # Update project references in tsconfig.json
      $ workspaces-to-typescript-project-references
      # Test on CI
      $ workspaces-to-typescript-project-references --check
`,
    {
        flags: {
            root: {
                type: "string",
                default: process.cwd()
            },
            check: {
                type: "boolean",
                default: false
            },
            plugin: {
                type: "string",
                isMultiple: true
            },
            tsconfigPath: {
                type: "string",
                default: DEFAULT_TSCONFIGPATH
            }
        },
        autoHelp: true,
        autoVersion: true
    }
);

export const run = async (
    _input = cli.input,
    flags = cli.flags
): Promise<{ exitStatus: number; stdout: string | null; stderr: string | null }> => {
    const plugins = Array.isArray(flags.plugin)
        ? flags.plugin.map((pluginPath) => {
              const plugin = require(path.join(process.cwd(), pluginPath));
              if (typeof plugin.plugin !== "function") {
                  throw new Error("plugin should export { plugin }.");
              }
              return plugin.plugin;
          })
        : undefined;
    const customTsConfigFinder = (location: string) => {
        return path.join(location, flags.tsconfigPath);
    };
    const result = toProjectReferences({
        rootDir: flags.root,
        checkOnly: flags.check,
        plugins,
        tsConfigPath: flags.tsconfigPath,
        tsConfigPathFinder: flags.tsconfigPath ? customTsConfigFinder : undefined
    });
    if (result.ok) {
        return {
            exitStatus: 0,
            stdout: flags.check ? "" : "Update Project References!",
            stderr: null
        };
    } else if (result.aggregateError) {
        return {
            exitStatus: 1,
            stdout: null,
            stderr: result.aggregateError.message + "\n\n" + result.aggregateError.errors.join("\n")
        };
    }
    return {
        exitStatus: 0,
        stdout: null,
        stderr: null
    };
};
