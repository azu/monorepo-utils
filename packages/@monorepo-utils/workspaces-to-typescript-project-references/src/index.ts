import fs from "fs";
import path from "upath";
import commentJSON from "comment-json";
import { plugin as workspacesPlugin } from "./manager/workspaces";
import assert from "assert";
import { PackageManagerPlugin } from "./manager/PackageManagerPlugin";

export const DEFAULT_TSCONFIGPATH = "tsconfig.json";

export type Options = {
    rootDir: string;
    checkOnly: boolean;
    plugins?: PackageManagerPlugin[];
    tsConfigPath?: string;
    tsConfigPathFinder?(location: string): string;
};
export type ToProjectReferencesResult =
    | {
          ok: true;
      }
    | {
          ok: false;
          aggregateError: {
              message: string;
              errors: Error[];
          };
      };
export const toProjectReferences = (options: Options) => {
    const plugins = Array.isArray(options.plugins) && options.plugins.length > 0 ? options.plugins : [workspacesPlugin];
    const pluginImplementations = plugins.map((plugin) => plugin(options));
    // use first plugin
    const supportPlugin = pluginImplementations.find((plugin) => {
        return plugin.supports();
    });
    if (!supportPlugin) {
        throw new Error("Not found supported plugin");
    }
    const relativeName = (pkgPath: string) => {
        return path.relative(options.rootDir, pkgPath);
    };
    const allPackages = supportPlugin.getAllPackages();
    const errors: Error[] = [];
    allPackages.forEach((packageInfo) => {
        const tsconfigFilePath =
            options.tsConfigPathFinder?.(packageInfo.location) ?? path.join(packageInfo.location, DEFAULT_TSCONFIGPATH);
        if (!fs.existsSync(tsconfigFilePath)) {
            // Skip has not tsconfig.json
            return;
        }
        const tsconfigJSON = commentJSON.parse(fs.readFileSync(tsconfigFilePath, "utf-8"));
        const references = supportPlugin.getDependencies(packageInfo.packageJSON);
        const newProjectReferences = references
            .map((reference) => {
                const absolutePathOrNull = supportPlugin.resolve(reference);
                if (!absolutePathOrNull) {
                    return;
                }
                const tsConfigDirName = path.dirname(options.tsConfigPath ?? DEFAULT_TSCONFIGPATH);
                const absolutePath = path.join(absolutePathOrNull, tsConfigDirName);
                if (!path.isAbsolute(absolutePath)) {
                    throw new Error(
                        `Plugin#resolve should return absolute path: ${absolutePath}, plugin: ${supportPlugin}`
                    );
                }
                if (packageInfo.location === absolutePath) {
                    const selfName = relativeName(packageInfo.location);
                    errors.push(
                        new Error(
                            `[${selfName}] Self dependencies is something wrong: ${selfName} refer to ${relativeName(
                                absolutePath
                            )}`
                        )
                    );
                }
                const resolvePath = path.dirname(path.resolve(packageInfo.location, tsconfigFilePath));
                return {
                    path: path.relative(resolvePath, absolutePath)
                };
            })
            .filter((r) => Boolean(r));
        const currentProjectReferences: { path: string }[] = tsconfigJSON["references"] ?? [];
        if (options.checkOnly) {
            // check
            try {
                // create pure object for compare
                const cleanCurrentProjectReferences = JSON.parse(
                    JSON.stringify(commentJSON.parse(commentJSON.stringify(currentProjectReferences), undefined, true))
                );
                assert.deepStrictEqual(cleanCurrentProjectReferences, newProjectReferences);
            } catch (error) {
                const selfName = relativeName(packageInfo.location);
                errors.push(new Error(`[${selfName}] ${error.message}`));
            }
            return;
        } else {
            // update
            if (errors.length === 0) {
                tsconfigJSON["references"] = newProjectReferences;
                fs.writeFileSync(tsconfigFilePath, commentJSON.stringify(tsconfigJSON, null, 2), "utf-8");
            }
        }
    });
    if (errors.length > 0) {
        return {
            ok: false,
            aggregateError: {
                message: `workspaces-to-typescript-project-references found ${errors.length} errors.

Please update your tsconfig.json via following command.

$ workspaces-to-typescript-project-references
`,
                errors
            }
        };
    }
    return {
        ok: true
    };
};
