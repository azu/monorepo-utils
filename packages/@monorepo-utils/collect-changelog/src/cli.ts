// MIT Â© 2018 azu
"use strict";

import * as fs from "fs";

import * as path from "path";
import { findChangelog } from "./collect-changelog-from-tag";
import Handlebars = require("handlebars");

export interface ExecuteOptions {
    directory: string;
    lernaTag: string;
    templatePath?: string;
}

export async function execute(options: ExecuteOptions) {
    const templatePath = options.templatePath
        ? path.resolve(process.cwd(), options.templatePath)
        : path.join(__dirname, "../template/changelog.handlebars");
    const code = fs.readFileSync(templatePath, "utf-8");
    const template = Handlebars.compile(code);
    const changelog = await findChangelog(options.directory, options.lernaTag);
    return template(changelog).trim();
}
