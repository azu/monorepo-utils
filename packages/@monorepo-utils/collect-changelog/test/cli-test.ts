import * as path from "path";
import { execute } from "../src/cli";

const fixturesDir = path.join(__dirname, "fixtures");
describe("cli", () => {
    it("textlint@10.2.0 should has changelog ", async () => {
        const actual = await execute({
            directory: fixturesDir,
            lernaTag: "textlint@10.2.0"
        });
        expect(actual).toMatchSnapshot();
    });
    it("@textlint/kernel@2.0.0-next. should has changelog ", async () => {
        const actual = await execute({
            directory: fixturesDir,
            lernaTag: "@textlint/kernel@2.0.0-next.0"
        });
        expect(actual).toMatchSnapshot();
    });
    it("@textlint/ast-tester@2.0.7 should be empty ", async () => {
        const actual = await execute({
            directory: fixturesDir,
            lernaTag: "@textlint/ast-tester@2.0.7"
        });
        expect(actual).toBe("");
    });
});
