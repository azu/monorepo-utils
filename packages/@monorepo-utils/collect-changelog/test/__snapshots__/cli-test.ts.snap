// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli @textlint/kernel@2.0.0-next. should has changelog  1`] = `
"## @textlint/kernel@2.0.0-next.0

### fixes

- **textlint:** fix kernel test ([#374](https://github.com/textlint/textlint/issues/374)) ([d6953cc](https://github.com/textlint/textlint/commit/d6953cc))
- **textlint-fixer-formatter:** use paths ([734806f](https://github.com/textlint/textlint/commit/734806f))"
`;

exports[`cli textlint@10.2.0 should has changelog  1`] = `
"## textlint@10.2.0

### fixes

- **textlint:** check textlintrc option exists for internal use ([744da23](https://github.com/textlint/textlint/commit/744da23))

### features

- **textlint:** support --no-textlintrc ([466e257](https://github.com/textlint/textlint/commit/466e257))
- **textlint:** add tests for --no-textlintrc ([289deb3](https://github.com/textlint/textlint/commit/289deb3))
- **textlint:** pass textlintrc options for failure tests ([bc0804d](https://github.com/textlint/textlint/commit/bc0804d))
- **textlint:** simplify textlint --help test ([0d2114d](https://github.com/textlint/textlint/commit/0d2114d))"
`;
